cob_base_velocity_smoother/src/cob_base_velocity_smoother/velocity_smoother.cpp

```diff
@@ -114,179 +113,120 @@ void VelocitySmoother::odometryCB(const nav_msgs::Odometry::ConstPtr& msg)
-      ROS_WARN("Using robot velocity feedback (%s) instead of last command: %f, %f, %f, %f",
-                robot_feedback == ODOMETRY ? "odometry" : "end commands",
-               (ros::Time::now()      - last_cb_time).toSec(),
-                current_vel.linear.x  - last_cmd_vel.linear.x,
-                current_vel.linear.y  - last_cmd_vel.linear.y,
-                current_vel.angular.z - last_cmd_vel.angular.z);
-      last_cmd_vel = current_vel;
-    }
+  double decel_vx = decel_factor*accel_vx;
+  double decel_vy = decel_factor*accel_vy;
+  double decel_w  = decel_factor*accel_w;
 
-    geometry_msgs::TwistPtr cmd_vel;
+  if ((robot_feedback != NONE) && (input_active == true) && (cb_avg_time > 0.0) &&
+      ((last_command > std::min(PERIOD_RECORD_SIZE*cb_avg_time, 0.5))    || //missed PERIOD_RECORD_SIZE input cycles or no input for 0.5 secs
+       (std::abs(current_vel.linear.x  - last_cmd_vel.linear.x)  > 0.2)  ||
+       (std::abs(current_vel.linear.y  - last_cmd_vel.linear.y)  > 0.2)  ||
+       (std::abs(current_vel.angular.z - last_cmd_vel.angular.z) > 2.0)))
+  {
+    // If the publisher has been inactive for a while, or if our current commanding differs a lot
+    // from robot velocity feedback, we cannot trust the former; rely on robot's feedback instead
+    // This can happen mainly due to preemption of current controller on velocity multiplexer.
+    // TODO: current command/feedback difference thresholds are arbitrary; they should somehow
+    // be proportional to max v and w...
+    ROS_WARN("Last command differs to much from robot velocity feedback (%s). Use feedback instead!",
+              robot_feedback == ODOMETRY ? "ODOMETRY" : "COMMAND");
+    last_cmd_vel = current_vel;
+  }
 
-    if ((target_vel.linear.x  != last_cmd_vel.linear.x) ||
-        (target_vel.linear.y  != last_cmd_vel.linear.y) ||
-        (target_vel.angular.z != last_cmd_vel.angular.z))
+  if ((target_vel.linear.x  != last_cmd_vel.linear.x) ||
+      (target_vel.linear.y  != last_cmd_vel.linear.y) ||
+      (target_vel.angular.z != last_cmd_vel.angular.z))
+  {
+    // Try to reach target velocity ensuring that we don't exceed the acceleration limits
+    geometry_msgs::Twist cmd_vel = target_vel;
+
+    double vx_inc, vy_inc, w_inc;
+    double ax, ay, aw;
+    double max_ax, max_ay, max_aw;
+    double actual_ax, actual_ay, actual_aw;
+
+    vx_inc = target_vel.linear.x - last_cmd_vel.linear.x;
+    ax = vx_inc/cb_avg_time;
+    max_ax = (ax > 0.0)?accel_vx:decel_vx;
+    //actual_ax = (std::abs(ax) > max_ax)?sign(ax)*max_ax:ax;
+
+    vy_inc = target_vel.linear.y - last_cmd_vel.linear.y;
+    ay = vy_inc/cb_avg_time;
+    max_ay = (ay > 0.0)?accel_vy:decel_vy;
+    //actual_ay = (std::abs(ay) > max_ay)?sign(ay)*max_ay:ay;
+
+    w_inc = target_vel.angular.z - last_cmd_vel.angular.z;
+    aw = w_inc/cb_avg_time;
+    max_aw = (aw > 0.0)?accel_w:decel_w;
+    //actual_aw = (std::abs(aw) > max_aw)?sign(aw)*max_aw:aw;
```

Comment:
I think your are missing something. As far as i understand your code you always apply the max accelerations. If we get an acceleration below our limits as input the max acceleration is still used...